sdks/java/core/src/main/java/org/apache/beam/sdk/io/DefaultFilenamePolicy.java

```diff
@@ -138,9 +223,40 @@ public ResourceId unwindowedFilename(ResourceId outputDirectory, Context context
 
   @Override
   public ResourceId windowedFilename(ResourceId outputDirectory,
-      WindowedContext c, String extension) {
-    throw new UnsupportedOperationException("There is no default policy for windowed file"
-        + " output. Please provide an explicit FilenamePolicy to generate filenames.");
+      WindowedContext context, String extension) {
+
+    final PaneInfo paneInfo = context.getPaneInfo();
+    long currentPaneIndex = (paneInfo == null ? -1L
+        : paneInfo.getIndex());
+    boolean firstPane = (paneInfo == null ? false : paneInfo.isFirst());
+    boolean lastPane = (paneInfo == null ? false : paneInfo.isLast());
+    String windowStr = windowToString(context.getWindow());
+
+    String templateToUse = shardTemplate;
+    if (!isWindowedTemplate(shardTemplate)){
+      templateToUse = shardTemplate + DEFAULT_WINDOWED_SHARED_TEMPLATE_SUFFIX;
+    }
+
+    String filename = constructName(prefix.get(), templateToUse, suffix,
+        context.getShardNumber(),
+        context.getNumShards(), currentPaneIndex, firstPane, lastPane, windowStr)
+        + extension;
+    return outputDirectory.resolve(filename, StandardResolveOptions.RESOLVE_FILE);
+  }
+
+  /*
+   * Since not all windows have toString() that is nice or is compatible to be a part of file name.
+   */
+  private String windowToString(BoundedWindow window) {
+    if (window instanceof GlobalWindow) {
+      return "GlobalWindow";
+    }
+    if (window instanceof IntervalWindow) {
+      IntervalWindow iw = (IntervalWindow) window;
+      return String.format("IntervalWindow-from-%s-to-%s", iw.start().toString(),
+          iw.end().toString());
```

Comment:
Two comments:
1. I think we want shorter strings here - maybe just <start>-<end>?

2. Please check to see what Instant.toString() returns. I'm afraid it might be a bit longer than we'd like, in which case we could use one of the joda DateTimeFormatter classes to format this.